File: 105.buildTree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        if (inorder.size() == 0) return NULL;
        int first = preorder[0];
        TreeNode* root = new TreeNode(first);

        vector<int> left;
        vector<int> right;
        int flag = 0;
        
        for (auto num:inorder) {
            if (num == first) {flag = 1; continue;}
            if (flag == 0) left.push_back(num);
            if (flag == 1) right.push_back(num);
        }

        preorder.erase(preorder.begin());

        root->left = buildTree(preorder, left);
        root->right = buildTree(preorder, right);

        return root;
    }
};

File: 105.buildTree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
 func buildTree(preorder []int, inorder []int) *TreeNode {
    if len(preorder) == 0 {
        return nil;
    }

    var root TreeNode;
    
    if len(preorder) == 1 {
        root.Val = preorder[0];
        return &root;
    }

    for i := range inorder { 
        if (preorder[0] == inorder[i]) {
            inLeft := inorder[0:i]
            inRight := inorder[i+1:]
            preLeft := preorder[1:i+1]
            preRight := preorder[i+1:]

            root.Val = preorder[0]
            root.Left = buildTree(preLeft, inLeft)
            root.Right = buildTree(preRight, inRight)
            break;
        }
    }

    return &root;
}